<h2 id="class-intro">Class Intro</h2>

<p><strong>Dr. Clay Small</strong></p>

<ul>
  <li>go from Big Data to knowledge about living systems</li>
  <li>data visualization and interpretation</li>
  <li>genomic studies workflow
    <ul>
      <li>study design</li>
      <li>molecular bio
        <ul>
          <li>isolate good stuff</li>
        </ul>
      </li>
      <li>simualted bio
        <ul>
          <li>in silico</li>
        </ul>
      </li>
      <li>raw data manipulation
        <ul>
          <li>archive, format, filter, etc</li>
        </ul>
      </li>
      <li>data explo and visualization
        <ul>
          <li>summary stats, assess assumptions, plots</li>
        </ul>
      </li>
      <li>inference
        <ul>
          <li>parameter estimation, hypothesis testing</li>
        </ul>
      </li>
      <li>interpretation
        <ul>
          <li>reach biological conclusions</li>
        </ul>
      </li>
    </ul>
  </li>
  <li>goals of these studies
    <ul>
      <li>descriptive
        <ul>
          <li>describe functional elements (genes)</li>
          <li>describe structural elements</li>
        </ul>
      </li>
      <li>associative
        <ul>
          <li>connect genomic variation with phenotype</li>
          <li>getting at the genotype-phenotype map</li>
          <li>GWAS, QTL mapping</li>
          <li>Manhattan plot</li>
        </ul>
      </li>
      <li>experimental/functional
        <ul>
          <li>understand how elements drive biology (perturbation)</li>
          <li>knockdown methods</li>
        </ul>
      </li>
      <li>predictive
        <ul>
          <li>use models built from huge multidimensional datasets to classify unknown phentoypes</li>
          <li>often through machine learning</li>
        </ul>
      </li>
      <li>comparative
        <ul>
          <li>describe how and why genomic elements differ among linneages</li>
        </ul>
      </li>
    </ul>
  </li>
</ul>

<h2 id="genome-biology-fundamentals">Genome Biology Fundamentals</h2>

<p><strong>Dr. Clay Small</strong></p>

<ul>
  <li>c-value enigma
    <ul>
      <li>weak relationship between complexity and genome size</li>
      <li>huge amount of repeats accounts for some of it</li>
      <li>tough for us to imagine an inefficient system</li>
    </ul>
  </li>
  <li>contentious: junk DNA or functional?
    <ul>
      <li>junk to 98% functional to unnecessary</li>
      <li>functional duplicates?</li>
    </ul>
  </li>
</ul>

<h2 id="public-access-db">Public Access DB</h2>

<p><strong>Dr. Clay Small</strong></p>

<ul>
  <li>Genbank is an aggregate of 3 redundant databases
    <ul>
      <li>NCBI, EMBL, DDBJ</li>
    </ul>
  </li>
</ul>

<h2 id="using-r-for-data-visualization">Using R for Data Visualization</h2>

<p><strong>Dr. Clay Small</strong></p>

<ul>
  <li>interpretation vs stats
    <ul>
      <li>effect size
        <ul>
          <li>fold-size</li>
          <li>any quantification of relationship b/t variables</li>
        </ul>
      </li>
      <li>hypothesis test
        <ul>
          <li>statement about the relationship b/t variable</li>
        </ul>
      </li>
    </ul>
  </li>
</ul>

<h2 id="plot-colors">Plot colors</h2>

<p><strong>Dr. Clay Small</strong></p>

<p>iwanthue
colorbrewer2</p>

<h2 id="phylogenetic-context-for-genomics">Phylogenetic Context for Genomics</h2>

<p><strong>Dr. Clay Small</strong></p>

<ul>
  <li>homology
    <ul>
      <li>speciation and duplication produce homologs</li>
    </ul>
  </li>
  <li>paralogs: gene duplication event</li>
  <li>orthologs: originate from speciation events</li>
  <li>xenologs acquired from HGT, appear to be orthologs</li>
</ul>

<h2 id="phylogenetics-cont">Phylogenetics, cont.</h2>

<p><strong>Dr. Clay Small</strong></p>

<p><strong>Multiple Sequence Alignment</strong></p>

<ul>
  <li>multiple heuristic solutions have been been developed</li>
  <li>progressive alignment is a solution
    <ol>
      <li>All pairwise alignments using Needleman-Wunsch</li>
      <li>Calculate distance matrix for all sequence pairs</li>
      <li>Cluster sequences according to distance matrix</li>
      <li>Align closest pair(s) using Needleman-Wunsch</li>
      <li>Collapse pairs to consensus sequences and iterate pairwise alignments un3l last sequence is aligned</li>
    </ol>
  </li>
  <li>popular aligners
    <ul>
      <li>Clustal, T-Coffee, MUSCLE, MAFFT</li>
    </ul>
  </li>
</ul>

<p><strong>Molecular Phylogenetics</strong></p>

<ul>
  <li>can represent trees in text form with the Newick format
    <ul>
      <li>((A1,B1),(B2,(A2a,A2b)));</li>
    </ul>
  </li>
  <li>cladogram–branch lengths aribitrary</li>
  <li>phylogram–branch lengths proportional to amount of molecular change</li>
  <li>chronogram–branch lengths proportional to actual time between divergence events</li>
  <li>including an outgroup gives orintation to the divergence</li>
  <li>Monophyletic groups, or clades, contain all the descendants of a single common ancestor</li>
  <li>Polyphyletic and paraphyletic groups do not contain all the descendant lineages</li>
  <li><strong><em>Distance-based methods</em></strong></li>
  <li>Calculate “distances” among sequences by countng up pairwise differences</li>
  <li>Use the distance matrix to iteratively cluster the most similar sequences and groups of sequences as they are formed</li>
  <li>Fast, but often wrong</li>
  <li>Examples: UPGMA, neighbor-joining</li>
  <li><strong><em>Character-based methods</em></strong></li>
  <li>Each “site” in the sequence alignment considered individually
    <ul>
      <li>Parsimony
        <ul>
          <li>Tree requiring the fewest overall changes is “most likely”</li>
        </ul>
      </li>
      <li>Model-based
        <ul>
          <li>Statistically evaluate how well trees fit models of molecular evolution</li>
          <li>Examples: Bayesian, Maximum Likelihood</li>
        </ul>
      </li>
    </ul>
  </li>
  <li>distance methods produce a phenogram–overall similarity, not necessarily evolutionary relationships
    <ul>
      <li>still useful b/c fast, good starting topologies, good for gene expression clustering (not sequence comparison)</li>
      <li>OTU may share a character state
        <ul>
          <li>derived–common ancestor</li>
          <li>ancestral–not useful, old similarity
            <ul>
              <li>iguanas &amp; humans have five toes, horse has one</li>
            </ul>
          </li>
          <li>homoplasy–independent evolution has caused the same character state in seperate linneages</li>
        </ul>
      </li>
    </ul>
  </li>
  <li>model-based methods account for homoplasy-producing processes (e.g. reversals), and rate heterogeneity parameters are included
<strong><em>Maximum Likelihood</em></strong></li>
  <li>how well specific parameter values explain the data</li>
  <li>Likelihood is a conditonal probability
    <ul>
      <li>The likelihood is the probability of obtaining the observed data given a model, with parameters of specific values.</li>
    </ul>
  </li>
  <li>bootstrap for support
    <ul>
      <li>resamples data randomly and produces tree with these subsets, n times</li>
    </ul>
  </li>
</ul>

<h2 id="transcriptome-normalization">Transcriptome Normalization</h2>

<p><strong>Dr. Clay Small</strong></p>

<ul>
  <li>normalization at the bench
    <ul>
      <li>high quality total RNA</li>
      <li>maximize RNA diversity/evenness
        <ul>
          <li>many tissue types</li>
          <li>early stages in dev</li>
          <li>duplex-specific nuclease normaliztion</li>
        </ul>
      </li>
      <li>minimize polymorphism
        <ul>
          <li>use inbred individual</li>
          <li>avoid pooling many individuals</li>
        </ul>
      </li>
    </ul>
  </li>
  <li>assembly challenges
    <ul>
      <li>genome assembly
        <ul>
          <li>low complexity/repetitive DNA
            <ul>
              <li>breaks in assembly</li>
            </ul>
          </li>
          <li>low coverage region
            <ul>
              <li>can’t connect nodes, low info</li>
              <li>breaks in assembly</li>
            </ul>
          </li>
          <li>high coverage regions sequencing error accumulation</li>
          <li>sequencing error and polymorphism
            <ul>
              <li>bubbles in the graph</li>
            </ul>
          </li>
        </ul>
      </li>
      <li>transcriptome assembly
        <ul>
          <li>low complexity/repetitive DNA not an issue</li>
          <li>uneven coverage among transcripts
            <ul>
              <li>diff to set universal parameters (like kmer size)</li>
              <li>sequencing errors accumulate in high-coverage areas</li>
            </ul>
          </li>
          <li>sequencing error and polymorphism
            <ul>
              <li>bubbles in the graph</li>
            </ul>
          </li>
        </ul>
      </li>
    </ul>
  </li>
  <li>kmer spectra
    <ul>
      <li>low end (low coverage) is sequencing error, SNP</li>
      <li>high end (high coverage) is repetitive elements</li>
      <li>measure of abundance on the y
        <ul>
          <li>number of kmers, density</li>
        </ul>
      </li>
    </ul>
  </li>
  <li>kmer filtering
    <ul>
      <li>genome assembly
        <ul>
          <li>filter out reads containing rare kmers</li>
          <li>filter out reads containing abundant kmers</li>
        </ul>
      </li>
      <li>transcriptome assembly
        <ul>
          <li>filter out reads for which a small subset of kmers are rare</li>
          <li>filter out reads containing abundant kmers to a point, but not completely</li>
        </ul>
      </li>
      <li>see kmer_filter from Stacks
        <ul>
          <li>discards reads for which at least 0.8*k consec kmers drop below median kmer coverage</li>
          <li>calculated from dataset</li>
        </ul>
      </li>
    </ul>
  </li>
</ul>

<h2 id="genome-annotation">Genome Annotation</h2>

<p><strong>Alison Fuiten</strong></p>

<ul>
  <li>structural annotation</li>
  <li>functional annotation</li>
  <li>Maker is annotation software
    <ul>
      <li>product is .gff</li>
    </ul>
  </li>
  <li>source of annotations
    <ul>
      <li>RNA and protein evidence
        <ul>
          <li>transcripts, ESTs, cDNA proteomes of closely related species</li>
        </ul>
      </li>
      <li>ab initio computational evidence
        <ul>
          <li>gene signals
            <ul>
              <li>start and stop codons</li>
              <li>splice junctions</li>
            </ul>
          </li>
          <li>Hidden Markov model introns and exons
            <ul>
              <li>the state (intron, exon) is hidden</li>
            </ul>
          </li>
        </ul>
      </li>
    </ul>
  </li>
</ul>

<h2 id="genome-assembly-w-trinity">Genome Assembly w/ Trinity</h2>

<p><strong>Dr. Clay Small</strong></p>

<ul>
  <li>de Bruijn assembly
    <ul>
      <li>sequencing error, ploymorphism creates a bubble (two paths)</li>
    </ul>
  </li>
  <li>Trinity
    <ul>
      <li>inchworm
        <ul>
          <li>kmer processing</li>
          <li>produces contigs</li>
        </ul>
      </li>
      <li>chrysalis
        <ul>
          <li>uses contigs to construct de Bruijn graphs</li>
        </ul>
      </li>
      <li>butterfly
        <ul>
          <li>reconstructs plausible, full-length, linear transcripts by reconciling the individual de Bruijn graphs</li>
        </ul>
      </li>
    </ul>
  </li>
  <li>others
    <ul>
      <li>Oases</li>
      <li>velvet</li>
    </ul>
  </li>
  <li>quality of assemblies
    <ul>
      <li>intrinsic
        <ul>
          <li>mean, median, max., min. contig length (plot the length distribuIon)</li>
          <li>number of contigs (or number of “isogroups”)
            <ul>
              <li>e.g., we expect ~20,000 coding genes for vertebrate genomes</li>
            </ul>
          </li>
          <li>proportion of reads that remap uniquely
            <ul>
              <li>many multiply-mapped reads imply redundancy</li>
            </ul>
          </li>
          <li>Blast contigs against themselves
            <ul>
              <li>many high-scoring hits imply redundancy</li>
            </ul>
          </li>
        </ul>
      </li>
      <li>extrinsic
        <ul>
          <li>Blast against “reference” cDNA or protein sequences
            <ul>
              <li>high proportion of “many-to-one” hits implies fragmentation</li>
            </ul>
          </li>
          <li>if primarily interested in coding sequence:
            <ul>
              <li>Proportion of reference ORFs in contigs (expectation is close to 1.0 for complex transcriptomes)</li>
            </ul>
          </li>
        </ul>
      </li>
    </ul>
  </li>
</ul>

<h2 id="digital-gene-expression">Digital Gene Expression</h2>

<p><strong>Dr. Clay Small</strong></p>

<ul>
  <li>aligning to genome, have features in annotation
    <ul>
      <li>always missing some features, some error in annotation</li>
    </ul>
  </li>
</ul>

<ol>
  <li>demulitplex and clean reads
    <ul>
      <li>.fastq &gt; .fastq</li>
      <li>Stacks</li>
    </ul>
  </li>
  <li>build reference with splice-aware aligner
    <ul>
      <li>meed top level assembly</li>
      <li>need annotation file</li>
      <li>.fasta &amp; .gR &gt; gsnap indices</li>
      <li>ensembl</li>
    </ul>
  </li>
  <li>align reads form each sample to genome
    <ul>
      <li>.fastq &gt; .sam</li>
      <li>gmap-gsnap</li>
    </ul>
  </li>
  <li>count uniquely mapped reads per “feature” in each sample
    <ul>
      <li>htseq package, python</li>
      <li>.sam &amp; .gR &gt; .txt</li>
    </ul>
  </li>
  <li>normalize counts per feature across samples and test for diffs between sample groups
    <ul>
      <li>.txt &gt; dataframes, results, figs</li>
      <li>R, python</li>
    </ul>
  </li>
</ol>

<ul>
  <li>ht-seq outputs read counts per gene feature
    <ul>
      <li>counts based on –mode
        <ul>
          <li>most commonly used is union</li>
        </ul>
      </li>
      <li>use unix join to compile many samples</li>
      <li>then use r for diff analysis</li>
    </ul>
  </li>
</ul>

<h2 id="intro-to-differential-expression-analysis">Intro to Differential Expression Analysis</h2>

<p><strong>Dr. Clay Small</strong></p>

<ul>
  <li>RNAseq is a sampling process
    <ul>
      <li>many downsampling steps</li>
      <li>will never know actual expression</li>
      <li>product is counts based on gene models and gene annotation</li>
    </ul>
  </li>
  <li>RPKM (Reads per kilobase per million)
    <ul>
      <li>standardize by gene length</li>
      <li>standardize by library depth</li>
    </ul>
  </li>
  <li>normalizing count data
    <ul>
      <li>major among-individual differences for some transcripts can affect perceived differences for other transcripts</li>
      <li>solutions?
        <ul>
          <li>some advocate normalizing all read counts by “houskeeping” gene counts within a sample</li>
          <li>TMM (Trimmed mean of M-value normalization)</li>
        </ul>
      </li>
    </ul>
  </li>
  <li>Want to test whether the number of reads for gene X drawn from sample type A differs from the number drawn from type B
    <ul>
      <li>Poisson take mean, variance
        <ul>
          <li>high counts come close to a normal distribution</li>
        </ul>
      </li>
      <li>RNAseq data not Poisson distributed, higher variance compared to the expected mean for a given gene
        <ul>
          <li>Poisson expectation looks like linear relationship b/t log(var) log(mean)</li>
        </ul>
      </li>
    </ul>
  </li>
  <li>Negative Binomial distribution
    <ul>
      <li>better than Poisson</li>
      <li>Problem: Variance must be estimated for every gene, and RNA-Seq studies tend to have few replicates</li>
      <li>Solution: “Borrow” information about variance from the other genes to temper the genewise estimates</li>
    </ul>
  </li>
  <li>Continuous Distribution Approach</li>
</ul>

<h2 id="python-vs-r">Python vs R</h2>

<p><strong>Shawn O’Neil</strong></p>

<ul>
  <li>r and Python are deeply different</li>
  <li>r modifies a copy
    <ul>
      <li>functions work with pass-by-value/copy</li>
    </ul>
  </li>
</ul>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">nums</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">2</span><span class="p">,</span><span class="m">3</span><span class="p">,</span><span class="m">4</span><span class="p">)</span><span class="w">
</span><span class="n">numsb</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">nums</span><span class="w">
</span><span class="n">numsb</span><span class="p">[</span><span class="m">0</span><span class="p">]</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1000</span><span class="w">
</span><span class="n">print</span><span class="p">(</span><span class="n">numsb</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 1 2 3 4
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">print</span><span class="p">(</span><span class="n">nums</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 1 2 3 4
</code></pre>
</div>

<ul>
  <li>python modifies original
    <ul>
      <li>pass-by-reference</li>
    </ul>
  </li>
</ul>

<div class="language-python highlighter-rouge"><pre class="highlight"><code><span class="n">nums</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">]</span>
<span class="n">numsb</span> <span class="o">=</span> <span class="n">nums</span>
<span class="n">numsb</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1000</span>
<span class="k">print</span><span class="p">(</span><span class="n">numsb</span><span class="p">)</span>
<span class="k">print</span><span class="p">(</span><span class="n">nums</span><span class="p">)</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1000, 2, 3, 4]
## [1000, 2, 3, 4]
</code></pre>
</div>

<ul>
  <li>objects and classes
    <ul>
      <li>object: thing
        <ul>
          <li>constructor</li>
          <li>other methods</li>
          <li>variables</li>
        </ul>
      </li>
    </ul>
  </li>
  <li>collections of data (which represent something in the world) and methods (functions)
    <ul>
      <li>instance variables (set of variables per object)</li>
    </ul>
  </li>
</ul>

<div class="language-python highlighter-rouge"><pre class="highlight"><code><span class="c">#!/usr/local/bin/python3.6</span>

<span class="k">class</span> <span class="nc">Gene</span><span class="p">:</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">i</span><span class="p">,</span> <span class="n">seq</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="nb">id</span> <span class="o">=</span> <span class="n">i</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">seq</span> <span class="o">=</span> <span class="n">seq</span>
        <span class="k">print</span><span class="p">(</span><span class="s">"Hello, world. I'm a new gene object."</span><span class="p">)</span>
   
    <span class="k">def</span> <span class="nf">get_avg_gc</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="nb">sum</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">:</span>
            <span class="nb">sum</span> <span class="o">=</span> <span class="nb">sum</span> <span class="o">+</span> <span class="n">g</span><span class="o">.</span><span class="n">get_gc_count</span><span class="p">()</span>

        <span class="k">return</span> <span class="nb">sum</span><span class="o">/</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">)</span>

<span class="n">g</span> <span class="o">=</span> <span class="n">Gene</span><span class="p">(</span><span class="s">"Id34"</span><span class="p">,</span> <span class="s">"ATCAGTACGAT"</span><span class="p">)</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Hello, world. I'm a new gene object.
</code></pre>
</div>

<ul>
  <li>API: application programming interface
    <ul>
      <li>list of methods &amp; variables</li>
    </ul>
  </li>
</ul>

<h4 id="python">Python</h4>

<ul>
  <li>the object orientation of python results in memory expense</li>
  <li>python is already composed of objects, so adding custom objects doesn’t intorduce penalties</li>
  <li>object oriented program in python is pretty much like java</li>
</ul>

<h4 id="r">R</h4>

<ul>
  <li>r is very old!</li>
  <li>Advanced R</li>
  <li>S3 classes
    <ul>
      <li>oldest</li>
      <li>ubiquitous</li>
    </ul>
  </li>
  <li>S4 classes
    <ul>
      <li>like S3 plus some other stuff</li>
    </ul>
  </li>
  <li>reference classes
    <ul>
      <li>brand new</li>
      <li>like python</li>
    </ul>
  </li>
  <li>lists are great in r, can store heterogenous data w/ diff lengths
    <ul>
      <li>a dataframe is a list, w/ columns of same length</li>
    </ul>
  </li>
</ul>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">g</span><span class="m">1</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">list</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s2">"GO:12"</span><span class="p">,</span><span class="w"> </span><span class="s2">"GO:72"</span><span class="p">),</span><span class="w">
         </span><span class="s2">"ID34"</span><span class="p">,</span><span class="w">
         </span><span class="s2">"TAGACTGAATTGC"</span><span class="p">,</span><span class="w">
         </span><span class="nf">list</span><span class="p">(</span><span class="m">2</span><span class="p">,</span><span class="m">7</span><span class="p">))</span><span class="w">
</span><span class="n">g</span><span class="m">1</span><span class="p">[[</span><span class="m">1</span><span class="p">]]</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] "GO:12" "GO:72"
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">g</span><span class="m">1</span><span class="p">[[</span><span class="m">2</span><span class="p">]]</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] "ID34"
</code></pre>
</div>

<ul>
  <li>attributes allow sdtoring data with metadata</li>
</ul>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">samp</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">rnorm</span><span class="p">(</span><span class="m">100</span><span class="p">,</span><span class="w"> </span><span class="n">mean</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">sd</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w">
</span><span class="nf">attr</span><span class="p">(</span><span class="n">samp</span><span class="p">,</span><span class="w"> </span><span class="s2">"DistType"</span><span class="p">)</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Normal"</span><span class="w">

</span><span class="nf">attr</span><span class="p">(</span><span class="n">samp</span><span class="p">,</span><span class="w"> </span><span class="s2">"DistType"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] "Normal"
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="nf">attr</span><span class="p">(</span><span class="n">g</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="s2">"class"</span><span class="p">)</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Gene"</span><span class="w">
</span><span class="c1">#class(g1 = "Gene") has the same function
</span><span class="w">
</span><span class="nf">attr</span><span class="p">(</span><span class="n">g</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="s2">"class"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] "Gene"
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># input becomes list of heterogenous data, class is set to Gene
</span><span class="n">Gene</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">g</span><span class="p">,</span><span class="w"> </span><span class="n">i</span><span class="p">,</span><span class="w"> </span><span class="n">s</span><span class="p">)</span><span class="w"> </span><span class="p">{</span><span class="w">
    </span><span class="n">newg</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">list</span><span class="p">(</span><span class="n">id</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">i</span><span class="p">,</span><span class="w"> </span><span class="n">seq</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">s</span><span class="p">,</span><span class="w"> </span><span class="n">go_nums</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">g</span><span class="p">)</span><span class="w">
    </span><span class="nf">class</span><span class="p">(</span><span class="n">newg</span><span class="p">)</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Gene"</span><span class="w">
    </span><span class="nf">return</span><span class="p">(</span><span class="n">newg</span><span class="p">)</span><span class="w">
</span><span class="p">}</span><span class="w">

</span><span class="n">g</span><span class="m">1</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Gene</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s2">"GO:12"</span><span class="p">,</span><span class="w"> </span><span class="s2">"GO:72"</span><span class="p">),</span><span class="w">
         </span><span class="s2">"ID34"</span><span class="p">,</span><span class="w">
         </span><span class="s2">"TAGACTGAATTGC"</span><span class="p">)</span><span class="w">
</span><span class="n">g</span><span class="m">1</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## $id
## [1] "ID34"
## 
## $seq
## [1] "TAGACTGAATTGC"
## 
## $go_nums
## [1] "GO:12" "GO:72"
## 
## attr(,"class")
## [1] "Gene"
</code></pre>
</div>

<ul>
  <li>say you want len_bp function for ntds and for proteins
    <ul>
      <li>naming convention is len_bp.gene, len_bp.protein</li>
      <li>not class aware, band-aid</li>
      <li>we want it to be class awre, and to dispatch to the correct function!</li>
    </ul>
  </li>
  <li>generic function calls class specific functions (method) clled dispatching</li>
  <li>UseMethod(“len_bp”, x) is the solution</li>
  <li>methods() reports possible functions for the class</li>
</ul>

<h4 id="javascript">Javascript</h4>

<ul>
  <li>very similar mindset to python</li>
  <li>check out processing</li>
  <li>just has numbers</li>
  <li>pass by reference</li>
</ul>

<div class="language-javascript highlighter-rouge"><pre class="highlight"><code><span class="kd">var</span> <span class="nx">a</span> <span class="o">=</span> <span class="mi">5</span><span class="p">;</span>
<span class="nx">myList</span> <span class="o">=</span> <span class="p">[];</span>
<span class="nx">myList</span><span class="p">.</span><span class="nx">push</span><span class="p">(</span><span class="s2">"joe"</span><span class="p">);</span>

<span class="err">#</span><span class="k">for</span> <span class="nx">el</span> <span class="k">in</span> <span class="nx">some</span><span class="p">.</span><span class="nx">list</span><span class="p">;</span>
<span class="err">#</span>   <span class="nx">print</span><span class="p">(</span><span class="nx">el</span><span class="p">)</span>
<span class="k">for</span><span class="p">(</span><span class="kd">var</span> <span class="nx">i</span> <span class="o">=</span> <span class="mi">0</span><span class="p">;</span> <span class="nx">i</span> <span class="o">&lt;</span> <span class="nx">someList</span><span class="p">.</span><span class="nx">length</span><span class="p">;</span> <span class="nx">i</span><span class="o">++</span><span class="p">)</span>

<span class="err">#</span><span class="nx">print</span><span class="p">(</span><span class="s2">"hi!"</span><span class="p">)</span>
<span class="nx">console</span><span class="p">.</span><span class="nx">log</span><span class="p">(</span><span class="s2">"hi"</span><span class="p">)</span>

<span class="nx">add_two_nums</span> <span class="o">=</span> <span class="nx">functions</span><span class="p">(</span><span class="nx">a</span><span class="p">,</span> <span class="nx">b</span><span class="p">)</span> <span class="p">{</span>
    <span class="kd">var</span> <span class="nx">sum</span> <span class="o">=</span> <span class="nx">a</span> <span class="o">+</span> <span class="nx">b</span><span class="p">;</span>
    <span class="k">return</span> <span class="nx">sum</span><span class="p">;</span>
<span class="p">}</span>
</code></pre>
</div>

<ul>
  <li>check out <a href="https://alpha.editor.p5js.org/">https://alpha.editor.p5js.org/</a></li>
  <li>check out <a href="https://www.openprocessing.org/class/56462">https://www.openprocessing.org/class/56462</a></li>
  <li>check out Daniel Shiffman</li>
</ul>

<div class="language-javascript highlighter-rouge"><pre class="highlight"><code><span class="kd">function</span> <span class="nx">Ball</span><span class="p">(</span><span class="nx">xloc</span><span class="p">,</span> <span class="nx">yloc</span><span class="p">,</span> <span class="nx">sx</span><span class="p">,</span> <span class="nx">sy</span><span class="p">,</span> <span class="nx">diameter</span><span class="p">)</span> <span class="p">{</span>
    <span class="k">this</span><span class="p">.</span><span class="nx">xloc</span> <span class="o">=</span> <span class="nx">xloc</span><span class="p">;</span>
    <span class="k">this</span><span class="p">.</span><span class="nx">yloc</span> <span class="o">=</span> <span class="nx">yloc</span><span class="p">;</span>
    <span class="k">this</span><span class="p">.</span><span class="nx">speedx</span> <span class="o">=</span> <span class="nx">sx</span><span class="p">;</span>
    <span class="k">this</span><span class="p">.</span><span class="nx">speedy</span> <span class="o">=</span> <span class="nx">sy</span><span class="p">;</span>
    <span class="k">this</span><span class="p">.</span><span class="nx">radius</span> <span class="o">=</span> <span class="nx">diameter</span><span class="p">;</span>

    <span class="k">this</span><span class="p">.</span><span class="nx">move</span> <span class="o">=</span> <span class="kd">function</span><span class="p">()</span> <span class="p">{</span>
        <span class="k">this</span><span class="p">.</span><span class="nx">xloc</span> <span class="o">=</span> <span class="k">this</span><span class="p">.</span><span class="nx">xloc</span> <span class="o">+</span> <span class="k">this</span><span class="p">.</span><span class="nx">speedx</span><span class="p">;</span>
        <span class="k">this</span><span class="p">.</span><span class="nx">yloc</span> <span class="o">=</span> <span class="k">this</span><span class="p">.</span><span class="nx">yloc</span> <span class="o">+</span> <span class="k">this</span><span class="p">.</span><span class="nx">speedy</span><span class="p">;</span>
    <span class="p">}</span>

    <span class="k">this</span><span class="p">.</span><span class="nx">display</span> <span class="o">=</span> <span class="kd">function</span><span class="p">()</span> <span class="p">{</span>
        <span class="nx">ellipse</span><span class="p">(</span><span class="k">this</span><span class="p">.</span><span class="nx">xloc</span><span class="p">,</span> <span class="k">this</span><span class="p">.</span><span class="nx">yloc</span><span class="p">,</span> <span class="k">this</span><span class="p">.</span><span class="nx">diameter</span><span class="p">,</span> <span class="k">this</span><span class="p">.</span><span class="nx">diameter</span><span class="p">);</span>
    <span class="p">}</span>
<span class="p">}</span>

<span class="kd">var</span> <span class="nx">b1</span> <span class="o">=</span> <span class="k">new</span> <span class="nx">Ball</span><span class="p">(</span><span class="mi">50</span><span class="p">,</span> <span class="mi">50</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">20</span><span class="p">);</span>

<span class="kd">function</span> <span class="nx">setup</span><span class="p">()</span> <span class="p">{</span>
    <span class="nx">createCanvas</span><span class="p">(</span><span class="mi">800</span><span class="p">,</span> <span class="mi">800</span><span class="p">);</span>
<span class="p">}</span>

<span class="kd">function</span> <span class="nx">draw</span><span class="p">()</span> <span class="p">{</span>
    <span class="nx">background</span><span class="p">(</span><span class="mi">224</span><span class="p">,</span> <span class="mi">190</span><span class="p">,</span> <span class="mi">173</span><span class="p">);</span>
    <span class="nx">b1</span><span class="p">.</span><span class="nx">display</span><span class="p">();</span>
    <span class="nx">b1</span><span class="p">.</span><span class="nx">move</span><span class="p">();</span>
<span class="p">}</span>
</code></pre>
</div>

<h2 id="multivariate-visualization">Multivariate Visualization</h2>

<p><strong>Dr. Clay Small</strong></p>

<h4 id="gsea">GSEA</h4>

<ul>
  <li>narrow the list to test <em>a priori</em></li>
  <li>gene set enrichment analysis (GSEA)
    <ul>
      <li>do groups of genes demonstrate patterns of differential expression more than the + transcriptome wide average</li>
      <li>groups defined <em>a priori</em></li>
      <li>groups defined by bio db mapping
        <ul>
          <li>KEGG pathways
            <ul>
              <li>the pathway is the set</li>
            </ul>
          </li>
          <li>gene ontology (GO) categories</li>
        </ul>
      </li>
      <li>GSEA isn’t limited to gene expression data</li>
    </ul>
  </li>
  <li>GSEA looks at fold change group vs group</li>
</ul>

<h4 id="multivariate-visualization-1">multivariate visualization</h4>

<ul>
  <li>start with correlation matrix</li>
  <li>from here do PCA</li>
  <li>non-metric multidimensional scaling (nMDS)</li>
  <li>heatmaps
    <ul>
      <li>reflect relativized gene expression levels for ind genes in ind samples</li>
      <li>want to organize genes and samples by similarity</li>
    </ul>
  </li>
</ul>
